source/retryable-reads/tests/README.rst

```diff
@@ -0,0 +1,197 @@
+  point may trigger before transitioning to ``"off"``.
+- ``{ skip: <integer> }`` may be used to defer the first trigger of a fail
+  point, after which it will transition to ``"alwaysOn"``.
+
+The ``data`` option is a document that may be used to specify options that
+control the fail point's behavior. ``failCommand`` supports the following
+``data`` options, which may be combined if desired:
+
+- ``failCommands``: Required, the list of command names to fail.
+- ``closeConnection``: Boolean option, which defaults to ``false``. If
+  ``true``, the connection on which the command is executed will be closed
+  and the client will see a network error.
+- ``errorCode``: Integer option, which is unset by default. If set, the
+  specified command error code will be returned as a command error.
+
+Disabling Fail Point after Test Execution
+-----------------------------------------
+
+After each test that configures a fail point, drivers should disable the
+``failCommand`` fail point to avoid spurious failures in
+subsequent tests. The fail point may be disabled like so::
+
+    db.runCommand({
+        configureFailPoint: "failCommand",
+        mode: "off"
+    });
+
+Network Error Tests
+===================
+
+Network error tests are expressed in YAML and should be run against a standalone,
+shard cluster, or single-node replica set. Until `SERVER-34943`_ is resolved,
+these tests cannot be run against a multi-node replicaset.
+
+.. _SERVER-34943: https://jira.mongodb.org/browse/SERVER-34943
+
+Test Format
+-----------
+
+Each YAML file has the following keys:
+
+- ``database_name`` and ``collection_name``: Optional. The database and collection to use
+  for testing.
+
+- ``data``: The data that should exist in the collection(s) under test before
+  each test run.
+
+- ``tests``: An array of tests that are to be run independently of each other.
+  Each test will have some or all of the following fields:
+
+  - ``description``: The name of the test.
+
+  - ``failPoint``: Optional, a server failpoint to enable expressed as the
+    configureFailPoint command to run on the admin database.
+
+  - ``operation``: A document describing an operation to be
+    executed. The document has the following fields:
+
+    - ``name``: The name of the operation on ``object``.
+
+    - ``object``: The name of the object to perform the operation on. Can be
+      "database", "collection", or "client."
+
+    - ``arguments``: Optional, the names and values of arguments.
+
+    - ``result``: Optional. The return value from the operation, if any. This
+      field may be a scalar (e.g. in the case of a count), a single document, or
+      an array of documents in the case of a multi-document read. If the
+      operation is expected to return an error, the ``result`` is a single
+      document that has one or more of the following fields:
+      - ``error``: Optional. If ``true``, the test should expect an error or
```

Comment:
I'm fine with `result.error`. I like the similarity to `result.errorCode` et al.
> As a minor plus for leaving error as a subefield of result, as it currently stands, the transactions test runner can be used to run these tests with very few modifications and (thus far) no conflicting changes

That's what I've done in Python. The major change is that these tests have a single `"operation"` while the transaction tests use a list of `"operations"`. The crud tests also use a single `"operation"` so I think that's fine too as long as we never need to run multiple operations in a single test.

Reply:
I disagree with nesting `error` under `result`. In the case of CRUD tests, an operation may raise an exception (`error: true`) but we still want to assert the BulkWriteResult included in the exception. In that case, `result` is a separate document with `insertedCount`, etc.

I imagine this doesn't come up at all in the retryable reads spec tests, since you don't deal with bulk write results, but it should explain why CRUD placed `error` as its own field parallel to `result`.

Reply:
I'm okay with `error` appearing at the same level as `result` if it helps future tests, even if retryable-reads admittedly doesn't need it.